# Synthetic Data Generation Using RAGAS - RAG Evaluation with LangSmith

In the following notebook we'll explore a use-case for RAGAS' synthetic testset generation workflow!



- 🤝 BREAKOUT ROOM #1
  1. Use RAGAS to Generate Synthetic Data

- 🤝 BREAKOUT ROOM #2
  1. Load them into a LangSmith Dataset
  2. Evaluate our RAG chain against the synthetic test data
  3. Make changes to our pipeline
  4. Evaluate the modified pipeline

SDG is a critical piece of the puzzle, especially for early iteration! Without it, it would not be nearly as easy to get high quality early signal for our application's performance.

Let's dive in!

# 🤝 BREAKOUT ROOM #1

## Task 1: Dependencies and API Keys

We'll need to install a number of API keys and dependencies, since we'll be leveraging a number of great technologies for this pipeline!

1. OpenAI's endpoints to handle the Synthetic Data Generation
2. OpenAI's Endpoints for our RAG pipeline and LangSmith evaluation
3. QDrant as our vectorstore
4. LangSmith for our evaluation coordinator!

Let's install and provide all the required information below!

## Dependencies and API Keys:

> NOTE: DO NOT RUN THESE CELLS IF YOU ARE RUNNING THIS NOTEBOOK LOCALLY

In [ ]:
#!pip install -qU ragas==0.2.10

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 175.7/175.7 kB 8.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.5/45.5 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 71.1/71.1 kB 5.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 480.6/480.6 kB 24.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.5/2.5 MB 68.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 48.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 411.6/411.6 kB 27.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.9/50.9 kB 3.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 454.8/454.8 kB 28.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 50.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 116.3/116.3 kB 8.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 179.3/179.3 kB 14.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 134.8/1

In [ ]:
#!pip install -qU langchain-community==0.3.14 langchain-openai==0.2.14 unstructured==0.16.12 langgraph==0.2.61 langchain-qdrant==0.2.0

### NLTK Import

To prevent errors that may occur based on OS - we'll import NLTK and download the needed packages to ensure correct handling of data.

In [1]:
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package punkt to /Users/debanshu/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /Users/debanshu/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


True

In [2]:
import os
import getpass

os.environ["LANGCHAIN_TRACING_V2"] = "true"
os.environ["LANGCHAIN_API_KEY"] = getpass.getpass("LangChain API Key:")

We'll also want to set a project name to make things easier for ourselves.

In [3]:
from uuid import uuid4

os.environ["LANGCHAIN_PROJECT"] = f"AIM - SDG - {uuid4().hex[0:8]}"

OpenAI's API Key!

In [4]:
os.environ["OPENAI_API_KEY"] = getpass.getpass("OpenAI API Key:")

## Generating Synthetic Test Data

We wil be using Ragas to build out a set of synthetic test questions, references, and reference contexts. This is useful because it will allow us to find out how our system is performing.

> NOTE: Ragas is best suited for finding *directional* changes in your LLM-based systems. The absolute scores aren't comparable in a vacuum.

### Data Preparation

We'll prepare our data - which should hopefull be familiar at this point since it's our Loan Data use-case!

Next, let's load our data into a familiar LangChain format using the `DirectoryLoader`.

In [5]:
from langchain_community.document_loaders import DirectoryLoader
from langchain_community.document_loaders import PyMuPDFLoader


path = "data/"
loader = DirectoryLoader(path, glob="*.pdf", loader_cls=PyMuPDFLoader)
docs = loader.load()

### Knowledge Graph Based Synthetic Generation

Ragas uses a knowledge graph based approach to create data. This is extremely useful as it allows us to create complex queries rather simply. The additional testset complexity allows us to evaluate larger problems more effectively, as systems tend to be very strong on simple evaluation tasks.

Let's start by defining our `generator_llm` (which will generate our questions, summaries, and more), and our `generator_embeddings` which will be useful in building our graph.

### Unrolled SDG

In [6]:
from ragas.llms import LangchainLLMWrapper
from ragas.embeddings import LangchainEmbeddingsWrapper
from langchain_openai import ChatOpenAI
from langchain_openai import OpenAIEmbeddings
generator_llm = LangchainLLMWrapper(ChatOpenAI(model="gpt-4.1-nano"))
generator_embeddings = LangchainEmbeddingsWrapper(OpenAIEmbeddings())

/Users/debanshu/Documents/AIE07/07_Synthetic_Data_Generation_and_LangSmith/.venv/lib/python3.13/site-packages/pysbd/segmenter.py:66: SyntaxWarning: invalid escape sequence '\s'
  for match in re.finditer('{0}\s*'.format(re.escape(sent)), self.original_text):
/Users/debanshu/Documents/AIE07/07_Synthetic_Data_Generation_and_LangSmith/.venv/lib/python3.13/site-packages/pysbd/lang/arabic.py:29: SyntaxWarning: invalid escape sequence '\.'
  txt = re.sub('(?<={0})\.'.format(am), '∯', txt)
/Users/debanshu/Documents/AIE07/07_Synthetic_Data_Generation_and_LangSmith/.venv/lib/python3.13/site-packages/pysbd/lang/persian.py:29: SyntaxWarning: invalid escape sequence '\.'
  txt = re.sub('(?<={0})\.'.format(am), '∯', txt)


Next, we're going to instantiate our Knowledge Graph.

This graph will contain N number of nodes that have M number of relationships. These nodes and relationships (AKA "edges") will define our knowledge graph and be used later to construct relevant questions and responses.

In [7]:
from ragas.testset.graph import KnowledgeGraph

kg = KnowledgeGraph()
kg

KnowledgeGraph(nodes: 0, relationships: 0)

The first step we're going to take is to simply insert each of our full documents into the graph. This will provide a base that we can apply transformations to.

In [8]:
from ragas.testset.graph import Node, NodeType

### NOTICE: We're using a subset of the data for this example - this is to keep costs/time down.
for doc in docs[:20]:
    kg.nodes.append(
        Node(
            type=NodeType.DOCUMENT,
            properties={"page_content": doc.page_content, "document_metadata": doc.metadata}
        )
    )
kg

KnowledgeGraph(nodes: 20, relationships: 0)

Now, we'll apply the *default* transformations to our knowledge graph. This will take the nodes currently on the graph and transform them based on a set of [default transformations](https://docs.ragas.io/en/latest/references/transforms/#ragas.testset.transforms.default_transforms).

These default transformations are dependent on the corpus length, in our case:

- Producing Summaries -> produces summaries of the documents
- Extracting Headlines -> finding the overall headline for the document
- Theme Extractor -> extracts broad themes about the documents

It then uses cosine-similarity and heuristics between the embeddings of the above transformations to construct relationships between the nodes.

In [9]:
from ragas.testset.transforms import default_transforms, apply_transforms

transformer_llm = generator_llm
embedding_model = generator_embeddings

default_transforms = default_transforms(documents=docs, llm=transformer_llm, embedding_model=embedding_model)
apply_transforms(kg, default_transforms)
kg

Applying HeadlinesExtractor:   0%|          | 0/17 [00:00<?, ?it/s]

Applying HeadlineSplitter:   0%|          | 0/20 [00:00<?, ?it/s]

unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node


Applying SummaryExtractor:   0%|          | 0/31 [00:00<?, ?it/s]

Property 'summary' already exists in node '4402b2'. Skipping!
Property 'summary' already exists in node 'c6cb01'. Skipping!
Property 'summary' already exists in node '115e72'. Skipping!
Property 'summary' already exists in node 'b59dd4'. Skipping!
Property 'summary' already exists in node '05a60c'. Skipping!
Property 'summary' already exists in node '9e45e8'. Skipping!
Property 'summary' already exists in node 'f29645'. Skipping!
Property 'summary' already exists in node 'a21de0'. Skipping!
Property 'summary' already exists in node '8f52f5'. Skipping!
Property 'summary' already exists in node '3cfb00'. Skipping!
Property 'summary' already exists in node '23b1e8'. Skipping!
Property 'summary' already exists in node 'db09d8'. Skipping!
Property 'summary' already exists in node 'fefa64'. Skipping!
Property 'summary' already exists in node '56289f'. Skipping!


Applying CustomNodeFilter:   0%|          | 0/6 [00:00<?, ?it/s]

Applying [EmbeddingExtractor, ThemesExtractor, NERExtractor]:   0%|          | 0/41 [00:00<?, ?it/s]

Property 'summary_embedding' already exists in node '115e72'. Skipping!
Property 'summary_embedding' already exists in node '05a60c'. Skipping!
Property 'summary_embedding' already exists in node 'c6cb01'. Skipping!
Property 'summary_embedding' already exists in node '4402b2'. Skipping!
Property 'summary_embedding' already exists in node '9e45e8'. Skipping!
Property 'summary_embedding' already exists in node '23b1e8'. Skipping!
Property 'summary_embedding' already exists in node 'f29645'. Skipping!
Property 'summary_embedding' already exists in node 'fefa64'. Skipping!
Property 'summary_embedding' already exists in node '56289f'. Skipping!
Property 'summary_embedding' already exists in node 'a21de0'. Skipping!
Property 'summary_embedding' already exists in node '8f52f5'. Skipping!
Property 'summary_embedding' already exists in node 'b59dd4'. Skipping!
Property 'summary_embedding' already exists in node '3cfb00'. Skipping!
Property 'summary_embedding' already exists in node 'db09d8'. Sk

Applying [CosineSimilarityBuilder, OverlapScoreBuilder]:   0%|          | 0/2 [00:00<?, ?it/s]

KnowledgeGraph(nodes: 39, relationships: 476)

We can save and load our knowledge graphs as follows.

In [10]:
kg.save("loan_data_kg.json")
loan_data_kg = KnowledgeGraph.load("loan_data_kg.json")
loan_data_kg

KnowledgeGraph(nodes: 39, relationships: 476)

Using our knowledge graph, we can construct a "test set generator" - which will allow us to create queries.

In [11]:
from ragas.testset import TestsetGenerator

generator = TestsetGenerator(llm=generator_llm, embedding_model=embedding_model, knowledge_graph=loan_data_kg)

However, we'd like to be able to define the kinds of queries we're generating - which is made simple by Ragas having pre-created a number of different "QuerySynthesizer"s.

Each of these Synthetsizers is going to tackle a separate kind of query which will be generated from a scenario and a persona.

In essence, Ragas will use an LLM to generate a persona of someone who would interact with the data - and then use a scenario to construct a question from that data and persona.

In [12]:
from ragas.testset.synthesizers import default_query_distribution, SingleHopSpecificQuerySynthesizer, MultiHopAbstractQuerySynthesizer, MultiHopSpecificQuerySynthesizer

query_distribution = [
        (SingleHopSpecificQuerySynthesizer(llm=generator_llm), 0.5),
        (MultiHopAbstractQuerySynthesizer(llm=generator_llm), 0.25),
        (MultiHopSpecificQuerySynthesizer(llm=generator_llm), 0.25),
]

#### ❓ Question #1:

What are the three types of query synthesizers doing? Describe each one in simple terms.

##### ✅ Answer:

### **1. SingleHopSpecificQuerySynthesizer (50% weight)**
**What it does**: Creates simple, direct questions that can be answered with information from a single document or chunk.
**In simple terms**: These are straightforward questions like "What is the maximum loan amount?" or "What are the eligibility requirements?" that require finding one specific piece of information.

### **2. MultiHopAbstractQuerySynthesizer (25% weight)**
**What it does**: Generates complex questions that require reasoning across multiple documents or concepts, but focus on abstract understanding rather than specific details.
**In simple terms**: These are questions that need you to think and connect ideas, like "How do the different loan programs compare in terms of student benefits?" or "What are the overall themes in the loan application process?"

### **3. MultiHopSpecificQuerySynthesizer (25% weight)**
**What it does**: Creates complex questions that require finding specific details from multiple sources and combining them.
**In simple terms**: These are detailed questions that need information from different places, like "What are the specific deadlines for Pell Grant applications and how do they relate to the academic calendar?" or "What are the exact income limits for different loan types and how do they affect eligibility?"

## **Why This Distribution Matters**

The 50/25/25 split ensures that:
- **50%** of test questions are simple and direct (testing basic retrieval)
- **25%** test abstract reasoning abilities
- **25%** test complex information synthesis

This creates a comprehensive test set that evaluates different aspects of RAG system performance, from simple fact-finding to complex reasoning and information integration.



Finally, we can use our `TestSetGenerator` to generate our testset!

In [13]:
testset = generator.generate(testset_size=10, query_distribution=query_distribution)
testset.to_pandas()

Generating personas:   0%|          | 0/3 [00:00<?, ?it/s]

Generating Scenarios:   0%|          | 0/3 [00:00<?, ?it/s]

Generating Samples:   0%|          | 0/11 [00:00<?, ?it/s]

user_input  \
0   Whitch department is resposible for academic p...   
1   What information does Volume 8 provide regardi...   
2   What is Title IV and how does it relate to pay...   
3   What is Volume 8 about in the context of finan...   
4   What is the significance of the Title IV in re...   
5   How does credit hour allocation for clinical e...   
6   clinical work in standard terms include non-te...   
7   nonstandard terms overlapping courses how defi...   
8   How do the requirements outlined in Volume 2 a...   
9   How does Volume 8 relate to the inclusion of c...   
10  How do Appendix A and Appendix B guide the dis...   

                                   reference_contexts  \
0   [Chapter 1 Academic Years, Academic Calendars,...   
1   [Inclusion of Clinical Work in a Standard Term...   
2   [Non-Term Characteristics A program that measu...   
3   [both the credit or clock hours and the weeks ...   
4   [Disbursement Timing in Subscription-Based Pro...   
5   [<1-hop>\n\nInclusion of Clinical Work in a St...   
6   [<1-hop>\n\nInclusion of Clinical Work in a St...   
7   [<1-hop>\n\nInclusion of Clinical Work in a St...   
8   [<1-hop>\n\nChapter 1 Academic Years, Academic...   
9   [<1-hop>\n\nInclusion of Clinical Work in a St...   
10  [<1-hop>\n\nDisbursement Timing in Subscriptio...   

                                            reference  \
0   The context does not specify which department ...   
1   Volume 8 states that clinical work conducted o...   
2   Title IV programs, except for the Federal Work...   
3   Volume 8 discusses the effect of accelerated p...   
4   Disbursement Timing in Subscription-Based Prog...   
5   The context explains that clinical work includ...   
6   Inclusion of clinical work in standard term pe...   
7   nonstandard terms are those not semesters, tri...   
8   Volume 2 details the academic year requirement...   
9   Volume 8 provides guidance on the inclusion of...   
10  Appendix B provides detailed guidance on disbu...   

                        synthesizer_name  
0   single_hop_specifc_query_synthesizer  
1   single_hop_specifc_query_synthesizer  
2   single_hop_specifc_query_synthesizer  
3   single_hop_specifc_query_synthesizer  
4   single_hop_specifc_query_synthesizer  
5   multi_hop_abstract_query_synthesizer  
6   multi_hop_abstract_query_synthesizer  
7   multi_hop_abstract_query_synthesizer  
8   multi_hop_specific_query_synthesizer  
9   multi_hop_specific_query_synthesizer  
10  multi_hop_specific_query_synthesizer

### Abstracted SDG

The above method is the full process - but we can shortcut that using the provided abstractions!

This will generate our knowledge graph under the hood, and will - from there - generate our personas and scenarios to construct our queries.



In [14]:
from ragas.testset import TestsetGenerator

generator = TestsetGenerator(llm=generator_llm, embedding_model=generator_embeddings)
dataset = generator.generate_with_langchain_docs(docs[:20], testset_size=10)

Applying HeadlinesExtractor:   0%|          | 0/17 [00:00<?, ?it/s]

Applying HeadlineSplitter:   0%|          | 0/20 [00:00<?, ?it/s]

unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node
unable to apply transformation: 'headlines' property not found in this node


Applying SummaryExtractor:   0%|          | 0/31 [00:00<?, ?it/s]

Property 'summary' already exists in node '4b45dc'. Skipping!
Property 'summary' already exists in node 'f57e1f'. Skipping!
Property 'summary' already exists in node 'b15108'. Skipping!
Property 'summary' already exists in node '8af4b3'. Skipping!
Property 'summary' already exists in node '1a267d'. Skipping!
Property 'summary' already exists in node '42846f'. Skipping!
Property 'summary' already exists in node '472c55'. Skipping!
Property 'summary' already exists in node '846f04'. Skipping!
Property 'summary' already exists in node '438338'. Skipping!
Property 'summary' already exists in node '677f27'. Skipping!
Property 'summary' already exists in node '08b397'. Skipping!
Property 'summary' already exists in node '7af484'. Skipping!
Property 'summary' already exists in node '6d2643'. Skipping!
Property 'summary' already exists in node '8f3536'. Skipping!


Applying CustomNodeFilter:   0%|          | 0/6 [00:00<?, ?it/s]

Applying [EmbeddingExtractor, ThemesExtractor, NERExtractor]:   0%|          | 0/41 [00:00<?, ?it/s]

Property 'summary_embedding' already exists in node 'f57e1f'. Skipping!
Property 'summary_embedding' already exists in node '472c55'. Skipping!
Property 'summary_embedding' already exists in node '4b45dc'. Skipping!
Property 'summary_embedding' already exists in node '438338'. Skipping!
Property 'summary_embedding' already exists in node '677f27'. Skipping!
Property 'summary_embedding' already exists in node 'b15108'. Skipping!
Property 'summary_embedding' already exists in node '42846f'. Skipping!
Property 'summary_embedding' already exists in node '1a267d'. Skipping!
Property 'summary_embedding' already exists in node '8af4b3'. Skipping!
Property 'summary_embedding' already exists in node '846f04'. Skipping!
Property 'summary_embedding' already exists in node '6d2643'. Skipping!
Property 'summary_embedding' already exists in node '08b397'. Skipping!
Property 'summary_embedding' already exists in node '7af484'. Skipping!
Property 'summary_embedding' already exists in node '8f3536'. Sk

Applying [CosineSimilarityBuilder, OverlapScoreBuilder]:   0%|          | 0/2 [00:00<?, ?it/s]

Generating personas:   0%|          | 0/3 [00:00<?, ?it/s]

Generating Scenarios:   0%|          | 0/3 [00:00<?, ?it/s]

Generating Samples:   0%|          | 0/12 [00:00<?, ?it/s]

In [21]:
dataset.to_pandas()

user_input  \
0   How does the FAFSA process incorporate the FAF...   
1   Can you tell me about the FAFSA and what happe...   
2   What is Federal Tax Information used for in FA...   
3   What is the link to the 2024-25 FAFSA Partner ...   
4   how returning FAFSA filers do they need to rea...   
5   How does the removal of COVID-19 related guida...   
6   How does the use of IRS data for FAFSA eligibi...   
7   How do restrictions on FAFAs initiating new ap...   
8   In chapter 2, student family size example and ...   
9   How do Chapters 2 and 5 relate to determining ...   
10  H0w does the FA-DDX impact the verifcation pro...   
11  How does the national guard status affect the ...   

                                   reference_contexts  \
0   [Application and Verification Guide Introducti...   
1   [Chapter 1: The Application Process We removed...   
2   [electronically due to limitations on access t...   
3   [Note: There will be two active portals that c...   
4   [<1-hop>\n\nApplication and Verification Guide...   
5   [<1-hop>\n\nApplication and Verification Guide...   
6   [<1-hop>\n\nApplication and Verification Guide...   
7   [<1-hop>\n\nelectronically due to limitations ...   
8   [<1-hop>\n\nStudent Citizenship (13) Family Si...   
9   [<1-hop>\n\nStudent Citizenship (13) Family Si...   
10  [<1-hop>\n\nApplication and Verification Guide...   
11  [<1-hop>\n\nveteran, or will be one by June 30...   

                                            reference  \
0   The FAFSA process has been overhauled by the F...   
1   Chapter 1 states that the Returning FAFSA File...   
2   The context does not explicitly explain what F...   
3   The link to the 2024-25 FAFSA Partner Portal i...   
4   The context indicates that the section on Retu...   
5   The removal of COVID-19 related guidance has l...   
6   The use of IRS data for FAFSA eligibility dete...   
7   Restrictions on FAFAs initiating new applicati...   
8   According to Chapter 2, the student can includ...   
9   Chapter 2 provides detailed examples of how fa...   
10  The FA-DDX, as described in the context, repla...   
11  Students who are members of the National Guard...   

                        synthesizer_name  
0   single_hop_specifc_query_synthesizer  
1   single_hop_specifc_query_synthesizer  
2   single_hop_specifc_query_synthesizer  
3   single_hop_specifc_query_synthesizer  
4   multi_hop_abstract_query_synthesizer  
5   multi_hop_abstract_query_synthesizer  
6   multi_hop_abstract_query_synthesizer  
7   multi_hop_abstract_query_synthesizer  
8   multi_hop_specific_query_synthesizer  
9   multi_hop_specific_query_synthesizer  
10  multi_hop_specific_query_synthesizer  
11  multi_hop_specific_query_synthesizer

We'll need to provide our LangSmith API key, and set tracing to "true".

# 🤝 BREAKOUT ROOM #2

## Task 4: LangSmith Dataset

Now we can move on to creating a dataset for LangSmith!

First, we'll need to create a dataset on LangSmith using the `Client`!

We'll name our Dataset to make it easy to work with later.

In [15]:
from langsmith import Client

client = Client()

dataset_name = "Loan Synthetic Data"

langsmith_dataset = client.create_dataset(
    dataset_name=dataset_name,
    description="Loan Synthetic Data"
)

We'll iterate through the RAGAS created dataframe - and add each example to our created dataset!

> NOTE: We need to conform the outputs to the expected format - which in this case is: `question` and `answer`.

In [16]:
for data_row in dataset.to_pandas().iterrows():
  client.create_example(
      inputs={
          "question": data_row[1]["user_input"]
      },
      outputs={
          "answer": data_row[1]["reference"]
      },
      metadata={
          "context": data_row[1]["reference_contexts"]
      },
      dataset_id=langsmith_dataset.id
  )

## Basic RAG Chain

Time for some RAG!


In [17]:
rag_documents = docs

To keep things simple, we'll just use LangChain's recursive character text splitter!


In [18]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size = 500,
    chunk_overlap = 50
)

rag_documents = text_splitter.split_documents(rag_documents)

We'll create our vectorstore using OpenAI's [`text-embedding-3-small`](https://platform.openai.com/docs/guides/embeddings/embedding-models) embedding model.

In [19]:
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(model="text-embedding-3-small")

As usual, we will power our RAG application with Qdrant!

In [20]:
from langchain_community.vectorstores import Qdrant

vectorstore = Qdrant.from_documents(
    documents=rag_documents,
    embedding=embeddings,
    location=":memory:",
    collection_name="Loan RAG"
)

In [21]:
retriever = vectorstore.as_retriever(search_kwargs={"k": 10})

To get the "A" in RAG, we'll provide a prompt.

In [22]:
from langchain.prompts import ChatPromptTemplate

RAG_PROMPT = """\
Given a provided context and question, you must answer the question based only on context.

If you cannot answer the question based on the context - you must say "I don't know".

Context: {context}
Question: {question}
"""

rag_prompt = ChatPromptTemplate.from_template(RAG_PROMPT)

For our LLM, we will be using TogetherAI's endpoints as well!

We're going to be using Meta Llama 3.1 70B Instruct Turbo - a powerful model which should get us powerful results!

In [23]:
from langchain_openai import ChatOpenAI

llm = ChatOpenAI(model="gpt-4.1-mini")

Finally, we can set-up our RAG LCEL chain!

In [24]:
from operator import itemgetter
from langchain_core.runnables import RunnablePassthrough, RunnableParallel
from langchain.schema import StrOutputParser

rag_chain = (
    {"context": itemgetter("question") | retriever, "question": itemgetter("question")}
    | rag_prompt | llm | StrOutputParser()
)

In [25]:
rag_chain.invoke({"question" : "What kinds of loans are available?"})

'The kinds of loans available include Direct Subsidized Loans, Direct Unsubsidized Loans, Direct PLUS Loans (including student Federal PLUS Loans and parent Direct PLUS Loans), Subsidized and Unsubsidized Federal Stafford Loans, and Federal SLS Loans. However, note that Subsidized and Unsubsidized Federal Stafford Loans, Federal SLS Loans, and Federal PLUS Loans were made under the Federal Family Education Loan (FFEL) Program before new FFEL loans ended on July 1, 2010. New loans are made under the Direct Loan Program, which includes Direct Subsidized Loans, Direct Unsubsidized Loans, and Direct PLUS Loans.'

## LangSmith Evaluation Set-up

We'll use OpenAI's GPT-4.1 as our evaluation LLM for our base Evaluators.

In [26]:
eval_llm = ChatOpenAI(model="gpt-4.1")

We'll be using a number of evaluators - from LangSmith provided evaluators, to a few custom evaluators!

In [27]:
from langsmith.evaluation import LangChainStringEvaluator, evaluate

qa_evaluator = LangChainStringEvaluator("qa", config={"llm" : eval_llm})

labeled_helpfulness_evaluator = LangChainStringEvaluator(
    "labeled_criteria",
    config={
        "criteria": {
            "helpfulness": (
                "Is this submission helpful to the user,"
                " taking into account the correct reference answer?"
            )
        },
        "llm" : eval_llm
    },
    prepare_data=lambda run, example: {
        "prediction": run.outputs["output"],
        "reference": example.outputs["answer"],
        "input": example.inputs["question"],
    }
)

empathy_evaluator = LangChainStringEvaluator(
    "criteria",
    config={
        "criteria": {
            "empathy": "Is this response empathetic? Does it make the user feel like they are being heard?",
        },
        "llm" : eval_llm
    }
)

#### 🏗️ Activity #2:

Highlight what each evaluator is evaluating.

- `qa_evaluator`:
- `labeled_helpfulness_evaluator`:
- `empathy_evaluator`:

##### ✅ Answer:

### **`qa_evaluator`:**
**What it evaluates**: Answer accuracy and correctness
**In simple terms**: This evaluator checks if the RAG system's answer is factually correct and matches the expected reference answer. It's like a teacher grading whether the student got the right answer to a test question.

### **`labeled_helpfulness_evaluator`:**
**What it evaluates**: How helpful and useful the response is to the user
**In simple terms**: This evaluator goes beyond just correctness - it assesses whether the answer actually helps the user solve their problem. It considers factors like completeness, clarity, and whether the response addresses what the user really needed to know, comparing it against the reference answer.

### **`empathy_evaluator`:**
**What it evaluates**: Whether the response is empathetic and makes the user feel heard
**In simple terms**: This evaluator checks if the RAG system responds in a caring, understanding way that makes the user feel like their concerns are being taken seriously. It's about the tone and emotional intelligence of the response, not just the factual content.

## **Why These Three Evaluators Matter**

Together, these evaluators provide a comprehensive assessment of RAG system performance:

1. **Accuracy** (qa_evaluator): Is the information correct?
2. **Utility** (labeled_helpfulness_evaluator): Is the response actually helpful?
3. **User Experience** (empathy_evaluator): Does the response feel caring and supportive?

This multi-dimensional evaluation ensures that the RAG system is not only factually accurate but also genuinely useful and user-friendly, which is especially important for sensitive topics like loan applications where users may be stressed or confused.


## LangSmith Evaluation

In [28]:
evaluate(
    rag_chain.invoke,
    data=dataset_name,
    evaluators=[
        qa_evaluator,
        labeled_helpfulness_evaluator,
        empathy_evaluator
    ],
    metadata={"revision_id": "default_chain_init"},
)

View the evaluation results for experiment: 'back-country-58' at:
https://smith.langchain.com/o/88c6730c-2899-4f8d-a2a1-13e382f99e7f/datasets/7a1176a1-53a4-4572-a2c8-bbdf6f3f3db4/compare?selectedSessions=44fc5ea8-73d3-4310-9ed9-14c9c342d68c




0it [00:00, ?it/s]

<ExperimentResults back-country-58>

## Dope-ifying Our Application

We'll be making a few changes to our RAG chain to increase its performance on our SDG evaluation test dataset!

- Include a "dope" prompt augmentation
- Use larger chunks
- Improve the retriever model to: `text-embedding-3-large`

Let's see how this changes our evaluation!

In [29]:
EMPATHY_RAG_PROMPT = """\
Given a provided context and question, you must answer the question based only on context.

If you cannot answer the question based on the context - you must say "I don't know".

You must answer the question using empathy and kindness, and make sure the user feels heard.

Context: {context}
Question: {question}
"""

empathy_rag_prompt = ChatPromptTemplate.from_template(EMPATHY_RAG_PROMPT)

In [30]:
rag_documents = docs

In [31]:
from langchain.text_splitter import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size = 1000,
    chunk_overlap = 50
)

rag_documents = text_splitter.split_documents(rag_documents)

#### ❓Question #2:

Why would modifying our chunk size modify the performance of our application?

##### ✅ Answer:

Modifying the chunk size can significantly impact RAG application performance in several ways:

### **Smaller Chunks (e.g., 500 characters)**
**Advantages:**
- **More precise retrieval**: Smaller chunks allow the system to find very specific information
- **Better for fact-finding**: Good for questions that need exact details or quotes
- **Reduced noise**: Less irrelevant information in each chunk
- **Faster processing**: Smaller chunks are quicker to embed and search

**Disadvantages:**
- **Context fragmentation**: Important context might be split across multiple chunks
- **Missing relationships**: Connections between related concepts might be lost
- **Incomplete answers**: The LLM might miss important context needed for comprehensive responses

### **Larger Chunks (e.g., 1000 characters)**
**Advantages:**
- **Better context preservation**: More complete information in each chunk
- **Improved coherence**: Related concepts stay together
- **Better for complex questions**: More context for reasoning and synthesis
- **Reduced chunk boundaries**: Less risk of cutting important information mid-sentence

**Disadvantages:**
- **More noise**: Might include irrelevant information
- **Less precise**: Harder to find specific details
- **Higher computational cost**: Larger chunks take more time to process

### **Real-World Impact**

In the notebook, changing from 500 to 1000 characters likely improved performance because:

1. **Loan documents are complex**: They contain interrelated information about eligibility, deadlines, amounts, and processes
2. **Questions require context**: Users often need to understand relationships between different loan aspects
3. **Better reasoning**: The LLM can make better connections when it has more complete context

### **Optimal Chunk Size Depends On:**
- **Document type**: Technical docs vs. narrative content
- **Question complexity**: Simple facts vs. complex reasoning
- **Content structure**: How information is organized in the source documents
- **Use case**: Whether you need precision or comprehensiveness

The key is finding the right balance between having enough context for good answers while maintaining precise retrieval capabilities.

In [32]:
from langchain_openai import OpenAIEmbeddings

embeddings = OpenAIEmbeddings(model="text-embedding-3-large")

#### ❓Question #3:

Why would modifying our embedding model modify the performance of our application?

##### ✅ Answer:

Modifying the embedding model can significantly impact RAG application performance in several key ways:

### **From `text-embedding-3-small` to `text-embedding-3-large`**

### **Performance Improvements:**

**1. Better Semantic Understanding**
- **Larger models** have more parameters and training data, leading to deeper understanding of language nuances
- **Better concept mapping**: More accurately captures relationships between related terms (e.g., "loan eligibility" vs "qualification requirements")
- **Improved context awareness**: Better at understanding domain-specific terminology in loan documents

**2. Enhanced Retrieval Accuracy**
- **More precise similarity matching**: Better at finding the most relevant chunks for a given question
- **Reduced false positives**: Less likely to retrieve irrelevant information
- **Better handling of paraphrasing**: Can match questions to content even when different words are used

**3. Improved Multilingual and Domain Performance**
- **Better domain adaptation**: More robust understanding of financial/loan terminology
- **Enhanced semantic similarity**: Better at recognizing when different phrases mean the same thing

### **Technical Differences:**

**`text-embedding-3-small`:**
- **Dimensions**: 1536
- **Speed**: Faster processing
- **Cost**: Lower computational cost
- **Accuracy**: Good for general use cases

**`text-embedding-3-large`:**
- **Dimensions**: 3072 (double the size)
- **Speed**: Slower processing
- **Cost**: Higher computational cost
- **Accuracy**: Superior semantic understanding

### **Real-World Impact in Loan Applications:**

**Better Matching Examples:**
- Question: "What are the income requirements?"
- `text-embedding-3-small` might miss chunks about "financial eligibility criteria" or "income thresholds"
- `text-embedding-3-large` would better understand these are the same concept

**Improved Complex Query Handling:**
- Question: "How do I apply for financial aid if I'm a first-generation student?"
- `text-embedding-3-large` would better connect concepts like "first-generation," "financial aid," "application process," and "special programs"

### **Trade-offs to Consider:**

**Advantages of Larger Model:**
- Higher retrieval accuracy
- Better semantic understanding
- More robust performance on complex queries

**Disadvantages:**
- Increased latency (slower response times)
- Higher API costs
- More computational resources required

### **When the Upgrade Matters Most:**

The performance improvement is most noticeable when:
- **Domain-specific terminology** is important (like loan/financial terms)
- **Complex, multi-concept questions** are common
- **Precision is critical** (avoiding irrelevant information)
- **Semantic similarity** is more important than exact keyword matching

In the context of loan applications, where users often ask complex questions requiring information from multiple parts of documents, the larger embedding model would provide significantly better retrieval performance, leading to more accurate and helpful responses.

In [33]:
vectorstore = Qdrant.from_documents(
    documents=rag_documents,
    embedding=embeddings,
    location=":memory:",
    collection_name="Loan Data for RAG"
)

In [34]:
retriever = vectorstore.as_retriever()

Setting up our new and improved DOPE RAG CHAIN.

In [35]:
empathy_rag_chain = (
    {"context": itemgetter("question") | retriever, "question": itemgetter("question")}
    | empathy_rag_prompt | llm | StrOutputParser()
)

Let's test it on the same output that we saw before.

In [36]:
empathy_rag_chain.invoke({"question" : "What kinds of loans are available?"})

"Thank you for your question. Based on the information shared in the context, it seems there are several types of loans available to students and their parents to help cover the cost of attendance:\n\n1. **Direct Subsidized Loans** – These are loans based on the student's financial need, where the government pays the interest while the student is in school. They have annual and aggregate limits depending on the student's year and dependency status.\n\n2. **Direct Unsubsidized Loans** – These loans are available regardless of financial need and the interest accumulates while the student is in school.\n\n3. **Direct PLUS Loans** – These are for parents of dependent students or for graduate/professional students themselves. There is no fixed limit on PLUS loans other than the total cost of attendance minus other aid.\n\nAdditionally, if a dependent student's parent is unable to obtain a Direct PLUS loan, the student may be eligible for additional Direct Unsubsidized Loan amounts up to ind

Finally, we can evaluate the new chain on the same test set!

In [37]:
evaluate(
    empathy_rag_chain.invoke,
    data=dataset_name,
    evaluators=[
        qa_evaluator,
        labeled_helpfulness_evaluator,
        empathy_evaluator
    ],
    metadata={"revision_id": "empathy_rag_chain"},
)

View the evaluation results for experiment: 'cold-cheek-66' at:
https://smith.langchain.com/o/88c6730c-2899-4f8d-a2a1-13e382f99e7f/datasets/7a1176a1-53a4-4572-a2c8-bbdf6f3f3db4/compare?selectedSessions=b16ca987-0920-4b29-af75-a6e70c838325




0it [00:00, ?it/s]

<ExperimentResults cold-cheek-66>

#### 🏗️ Activity #3:

Provide a screenshot of the difference between the two chains, and explain why you believe certain metrics changed in certain ways.


##### ✅ Answer:

![graph](./graph.png)
![output](./output.png)

### **Performance Comparison Summary:**

**Chain Names:**
- **back-country-58**: Default RAG chain (baseline)
- **cold-cheek-66**: Empathy-enhanced RAG chain (improved)

### **Key Metrics Analysis:**

#### **1. Correctness (QA Evaluator)**
**Change**: **Significant Decrease** 📉
**Results**: 
- Default Chain: 9/9 CORRECT (100%)
- Empathy Chain: 6/9 CORRECT (66.7%)
- **Change**: -33.3% decrease

**Why this changed**: The empathetic prompting caused the system to over-explain and add supportive language that didn't match the reference answers exactly.

#### **2. Empathy (Labeled Helpfulness Evaluator)**
**Change**: **Massive Improvement** 📈
**Results**:
- Default Chain: 0% empathetic responses
- Empathy Chain: 100% empathetic responses
- **Change**: +100% improvement

**Why this changed**: The empathy-enhanced prompt explicitly instructed the system to be supportive and understanding.

#### **3. Helpfulness (Labeled Helpfulness Evaluator)**
**Change**: **No Change** ➡️
**Results**:
- Default Chain: 60% helpful (6/10 responses)
- Empathy Chain: 60% helpful (6/10 responses)
- **Change**: 0% (identical performance)

**Why this stayed the same**: The evaluator's criteria for helpfulness were independent of the empathetic tone.

#### **4. Performance Metrics**
**Token Usage**: 3,194 → 1,821 tokens (-43% reduction)
**Cost**: $0.001334 → $0.0008556 (-36% reduction)
**Latency**: 1.42s → 2.17s (+53% increase)

### **Overall Assessment:**

The empathy enhancement achieved its primary goal of improving user experience through empathetic responses, while also reducing costs and token usage. However, this came at the cost of reduced factual accuracy, suggesting a trade-off between being supportive and being precise. The helpfulness metric remained unchanged, indicating that the evaluator's criteria for helpfulness were not influenced by the empathetic tone.